docs/compactor.md

```diff
@@ -0,0 +1,93 @@
+
+#### **How to swap full snapshot with compacted snapshot atomically**
+
+Currently, full snapshots and the subsequent delta snapshots are grouped under same prefix path in the snapstore. When a full snapshot is created, it is placed under a prefix/directory with the name comprising of timestamp. Then subsequent delta snapshots are also pushed into the same directory. Thus each prefix/directory contains a single full snapshot and the subsequent delta snapshots. So far, it is the job of ETCDBR to start main ETCD process and snapshotter process which takes full snapshot and delta snapshot periodically. But as per our proposal, compaction will be running as parallel process to main ETCD process and snapshotter process. So we can't reliably co-ordinate between the processes to achieve switching to the compacted snapshot as the base snapshot atomically.```
+
+**Current Directory Structure**
+```yaml
+- Backup-192345
+    - Full-Snapshot-0-1-192345
+    - Incremental-Snapshot-1-100-192355
+    - Incremental-Snapshot-100-200-192365
+    - Incremental-Snapshot-200-300-192375
+- Backup-192789
+    - Full-Snapshot-0-300-192789
+    - Incremental-Snapshot-300-400-192799
+    - Incremental-Snapshot-400-500-192809
+    - Incremental-Snapshot-500-600-192819
+```
+
+To solve the problem, proposal is:
+1. ETCDBR will take the first full snapshot after it starts main ETCD Process and snapshotter process. After taking the first full snapshot, snapshotter will continue taking full snapshots. On the other hand, ETCDBR compactor command will be run as periodic job in a separate pod and use the existing full or compacted snapshots to produce further compacted snapshots. Full snapshots and compacted snapshots will be named after same fashion. So, there is no need of any mechanism to choose which snapshots(among full and compacted snapshot) to consider as base snapshots. 
+2. Flatten the directory structure of backup folder. Save all the full snapshots, delta snapshots and compacted snapshots under same directory/prefix. Restorer will restore from full/compacted snapshots and delta snapshots sorted based on the revision numbers in name (or timestamp if the revision numbers are equal).
+
+**Proposed Directory Structure**
+```yaml
+Backup :
+    - Full-Snapshot-0-1-192355 (Taken by snapshotter)
+    - Incremental-Snapshot-revision-1-100-192365
+    - Incremental-Snapshot-revision-100-200-192375
+    - Full-Snapshot-revision-0-200-192379 (Taken by snapshotter)
+    - Incremental-Snapshot-revision-200-300-192385
+    - Full-Snapshot-revision-0-300-192386 (Taken by compaction job)
+    - Incremental-Snapshot-revision-300-400-192396
+    - Incremental-Snapshot-revision-400-500-192406
+    - Incremental-Snapshot-revision-500-600-192416
+    - Full-Snapshot-revision-0-600-192419 (Taken by snapshotter)
+    - Full-Snapshot-revision-0-600-192420 (Taken by compaction job)
+```
+
```

Comment:
```suggestion

##### What happens to the delta snapshots that were compacted?

The proposed `compaction` sub-command in `etcdbrctl` (and hence,  the `CronJob` provisioned by `etcd-druid` that will schedule it at a regular interval) would only upload the compacted full snapshot.
It will not delete the snapshots (delta or full snapshots) that were compacted.
These snapshots which were superseded by a freshly uploaded compacted snapshot would follow the same life-cycle as other older snapshots.
I.e. they will be garbage collected according to the configured backup snapshot retention policy.

For example, if an `exponential` retention policy is configured and if compaction is done every `30m` then there might be at most `48` additional (compacted) full snapshots (`24h * 2`) in the backup for the latest day. As time rolls forward to the next day, these additional compacted snapshots (along with the delta snapshots that were compacted into them) will get get garbage collected retaining only one full snapshot for the day before according to the retention policy.

```